Add Phansalkar local threshold algorithm

kleinerELM opened the pull request:
Applies Phansalkar local threshold to an array. Phansalkar is a modification of Sauvola technique to deal with low contrast images.

## Description

I need the Phansalkar thresholding filter for low contrast SEM images. I found an implementation in ImageJ, which worked great. But I did not find any mentioning of that filter for scikit. I used threshold_sauvola() as a template.

Source paper: 
Phansalskar N. et al. "Adaptive local thresholding for detection of nuclei in diversity stained cytology images.", International Conference on Communications and Signal Processing (ICCSP), pp. 218-220, 2011
DOI: 10.1109/ICCSP.2011.5739305

It should be noted, that the image has to be preprocessed using equalize_adapthist():

```
from skimage import data
from skimage.exposure import equalize_adapthist
image = data.moon()
image_eq = equalize_adapthist(image)
t_phansalkar = threshold_phansalkar(image_eq, window_size=15, k=0.25, p=2.0, q=10.0)
binary_image = image_eq > t_phansalkar
```


## Checklist

<!-- It's fine to submit PRs which are a work in progress! -->
<!-- But before they are merged, all PRs should provide: -->
- [x] [Docstrings for all functions](https://github.com/numpy/numpy/blob/master/doc/example.py)
- [ ] Benchmark in `./benchmarks`, if your changes aren't covered by an
  existing benchmark
- [x] Gallery example in `./doc/examples` (new features only)
- [x] Unit tests
- [x] Clean style in [the spirit of PEP8](https://www.python.org/dev/peps/pep-0008/)

<!-- For detailed information on these and other aspects see -->
<!-- the scikit-image contribution guidelines. -->
<!-- https://scikit-image.org/docs/dev/contribute.html -->

## For reviewers

<!-- Don't remove the checklist below. -->
- [x] Check that the PR title is short, concise, and will make sense 1 year
  later.
- [ ] Check that new functions are imported in corresponding `__init__.py`.
- [ ] Check that new features, API changes, and deprecations are mentioned in
      `doc/release/release_dev.rst`.

kleinerELM committed:
added threshold_phansalkar

Applies Phansalkar local threshold to an array. Phansalkar is a
modification of Sauvola technique to deal with low contrast images.

pep8speaks commented:
Hello @kleinerELM! Thanks for opening this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:

* In the file [`skimage/filters/thresholding.py`](https://github.com/scikit-image/scikit-image/blob/5b794716ea498c789a206a2fb2c7b7c3ea30d364/skimage/filters/thresholding.py):

> [Line 1222:72](https://github.com/scikit-image/scikit-image/blob/5b794716ea498c789a206a2fb2c7b7c3ea30d364/skimage/filters/thresholding.py#L1222): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 1232:80](https://github.com/scikit-image/scikit-image/blob/5b794716ea498c789a206a2fb2c7b7c3ea30d364/skimage/filters/thresholding.py#L1232): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (92 > 79 characters)
> [Line 1235:1](https://github.com/scikit-image/scikit-image/blob/5b794716ea498c789a206a2fb2c7b7c3ea30d364/skimage/filters/thresholding.py#L1235): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 1240:41](https://github.com/scikit-image/scikit-image/blob/5b794716ea498c789a206a2fb2c7b7c3ea30d364/skimage/filters/thresholding.py#L1240): [E202](https://duckduckgo.com/?q=pep8%20E202) whitespace before ')'
> [Line 1240:64](https://github.com/scikit-image/scikit-image/blob/5b794716ea498c789a206a2fb2c7b7c3ea30d364/skimage/filters/thresholding.py#L1240): [W292](https://duckduckgo.com/?q=pep8%20W292) no newline at end of file

kleinerELM committed:
fixed PEP 8 issues

removed whitespaces, vorrected line lengths and line endings

kleinerELM committed:
Fixed another PEP 8 annotation

lagru commented on skimage/filters/thresholding.py:
```diff
@@ -1177,3 +1178,63 @@ def threshold_multiotsu(image, classes=3, nbins=256):
+
+
+def threshold_phansalkar(image, window_size=15, k=0.25, r=None, p=2.0, q=10.0):
+    """Applies Phansalkar local threshold to an array. Phansalkar is a
+    modification of Sauvola technique to deal with low contrast images.
```
There is an empty line missing between the short and the long description here and between the section further below as well. You can find [a few examples here](https://www.sphinx-doc.org/en/master/usage/extensions/example_numpy.html) and in the [official reference](https://numpydoc.readthedocs.io/en/latest/) for the formatting we use throughout the library.

kleinerELM replied:
I hope i was able to resolve that issue. I added some empty lines for readabilty as sown in the examples.

lagru commented on skimage/filters/thresholding.py:
```diff
@@ -1177,3 +1178,63 @@ def threshold_multiotsu(image, classes=3, nbins=256):
+    -----
+    This algorithm is originally designed for detection of cell nuclei in low
+    contrast images. Therefore the historgram has to be equalized beforehand
+    using skimage.exposure.equalize_adapthist().
```
```suggestion
    using :func:`skimage.exposure.equalize_adapthist`.
```

lagru reviewed (commented):
Thank you for the contribution @kleinerELM! This looks promising and from a first look at the paper and your description this might fit with the other thresholding functions. I'll give this a more thorough review later.

kleinerELM committed:
Update skimage/filters/thresholding.py

format improvement

Co-authored-by: Lars Grüter <lagru@users.noreply.github.com>

kleinerELM committed:
improved formatting for threshold_phansalkar()

Added some empty lines for readability

rfezzani commented on skimage/filters/thresholding.py:
```diff
@@ -1177,3 +1178,69 @@ def threshold_multiotsu(image, classes=3, nbins=256):
+    >>> from skimage import data
+    >>> from skimage.exposure import equalize_adapthist
+    >>> image = data.moon()
+    >>> image_eq = equalize_adapthist(image)
```
Why not include this histogram equalization in the function since it seems that it is required for the method to work?

kleinerELM replied:
I just followed the threshold_sauvola() fuction (and most of the other thresholding functions as well). They all return a threshold mask instead of a binary image. (Is there actually any reason for that?)
Since image_eq is necessary to create the correct binary image at the end, the function should best output a binary image directly, or what do you think?

I can see one advantage not to include the histogram equalization into the thresholding function:
If you want to find the best parameters window_size, k, r, p, q for a special image set, it is much faster if you have to calculate equalize_adapthist() only once, instead every time you change a parameter.
For example:

```
def threshold_phansalkar_parameter_variation( image ):
    #equalize only once
    image_eq = equalize_adapthist(image)
    
    # list of parameters
    k_list = [0.2,0.25,0.3,0.35]
    p_list = [1.5,2.0,2.5,3.0]

    fig, ax = plt.subplots(len(p_list),len(k_list), figsize=(15, 6) )
    fig.subplots_adjust(hspace = .5)
    ax = ax.ravel()

    for j in range(len(p_list)):
        for i in range(len(k_list)):
            # calculate binary mask of thresh_phansalkar() without doing equalize_adapthist() every time
            thresh_phansalkar = threshold_phansalkar(image_eq, window_size=15, k=k_list[i], p=p_list[j], q=10.0)
            binary_phansalkar = image_eq > thresh_phansalkar

            ax[i+j*len(k_list)].imshow(binary_phansalkar, cmap=plt.cm.gray)
            ax[i+j*len(k_list)].set_title( "k = " + str( k_list[i] )+ ", p = " + str( (p_list[j]) ) )
            ax[i+j*len(k_list)].set_axis_off()

    plt.show()
```

If you think that this is too rare a usecase, then your suggestion would definitely be more user-friendly.

kleinerELM replied:
According to @jni the function should return a threshold map. Therefore, I thing it is best to leave the function. I documented the requirements of the function in the funtion itself and the example.

rfezzani commented on skimage/filters/thresholding.py:
```diff
@@ -1177,3 +1178,69 @@ def threshold_multiotsu(image, classes=3, nbins=256):
+
+    if r is None:
+        imin, imax = dtype_limits(image, clip_negative=False)
+        r = 0.5 * (imax - imin)
```
If the histogram equalization is integrated to the function, `imax - imin` can be replaced by `image.ptp()`

kleinerELM replied:
Thank you for that suggestion.
I tried it for some images. And actually r will allways be 1, if the imput image is a result of the histogram equalization.
Therefore, I would suggest skipping the calculation and set `r = 1.0` in that case.

What is the advantage of image.ptp()? I tried to benchmark the function and it seems image.ptp() was a little slower on my system than the current code. But that is negligible in comparison to the thresholding itself.

rfezzani reviewed (commented):
Thank you @kleinerELM for your contribution, I just left some comments ;)

jni commented on skimage/filters/thresholding.py:
```diff
@@ -1177,3 +1178,69 @@ def threshold_multiotsu(image, classes=3, nbins=256):
+
+
+def threshold_phansalkar(image, window_size=15, k=0.25, r=None, p=2.0, q=10.0):
+    """Applies Phansalkar local threshold to an array. Phansalkar is a
+    modification of Sauvola technique to deal with low contrast images.
```
@kleinerELM not yet fixed — the short description should fit in one line, followed by a blank line, followed by a longer explanation. In this case:

```suggestion
    """Find the Phansalkar local threshold for an array.
    
    Phansalkar is a modification of Sauvola thresholding designed to deal with
    low contrast images.
```

kleinerELM committed:
improved description of threshold_phansalkar()

Co-authored-by: Juan Nunez-Iglesias <juan.nunez-iglesias@monash.edu>

kleinerELM committed:
Modified a description and an error message

Added 'threshold_phansalkar' to the error message in _validate_window_size() and to the description of _mean_std()

kleinerELM committed:
Fixed another PEP 8 annotation

jni commented:
@kleinerELM btw this will need some tests, and I think it would be good to add to e.g. https://scikit-image.org/docs/dev/auto_examples/segmentation/plot_niblack_sauvola.html#sphx-glr-auto-examples-segmentation-plot-niblack-sauvola-py

It should definitely return a threshold map, not a binarised image.

kleinerELM committed:
Added threshold_phansalkar() to __init__

kleinerELM committed:
Added Phansalkar threshold to the examples

kleinerELM committed:
Fixed PEP 8 issues

kleinerELM commented:
> 
> 
> @kleinerELM btw this will need some tests, and I think it would be good to add to e.g. https://scikit-image.org/docs/dev/auto_examples/segmentation/plot_niblack_sauvola.html#sphx-glr-auto-examples-segmentation-plot-niblack-sauvola-py
> 
> It should definitely return a threshold map, not a binarised image.

Thank you for your patience. I added Phansalkar local threshold to the example and fixed the missing reference in skimage/filters/__init__.py .

While I was researching how to add the test functions I stumbled across the PR #2443 . It seems there was already a discussion including the Phansalkar threshold. Phansalkar was mistyped in that discussion, which is why I did not find that issue in the fist place.
Should I continue to work on the test suite or do we drop this PR? I personally would love to see this filter in Scikit.

kleinerELM committed:
added phansklar tests

kleinerELM committed:
fixed line length for PEP 8

kleinerELM commented:
@jni I added test functions according to similar filter functions and they seem to work properly.
Please let me know if there's anything more I can do.

kleinerELM committed:
extended try_all_threshold()

Added threshold_local(), threshold_niblack(), threshold_sauvola() and threshold_phansalkar() to the try_all_threshold() function. This may help selecting the best thresold filter for an image.

kleinerELM commented on skimage/filters/thresholding.py:
```diff
@@ -118,10 +122,22 @@ def try_all_threshold(image, figsize=(8, 5), verbose=True):
+            # standard case
+            def wrapper(im):
+                return im > func(im)
+
```
I don't know if you would accept this as good style, but it works for thresholding algorithms which require different input variables.

kleinerELM commented on skimage/filters/thresholding.py:
```diff
@@ -135,10 +151,14 @@ def wrapper(im):
+                           'Local': thresh(threshold_local),
+                           'Niblack': thresh(threshold_niblack),
+                           'Sauvola': thresh(threshold_sauvola),
+                           'Phansalkar': thresh(threshold_phansalkar)})
```
Even if Phansalkar will not be accepted, I think it would be helpfull, if the most available filters are included in this function.

kleinerELM commented on skimage/filters/thresholding.py:
```diff
@@ -118,10 +122,22 @@ def try_all_threshold(image, figsize=(8, 5), verbose=True):
         """
-        def wrapper(im):
-            return im > func(im)
+        if ( func.__name__ == 'threshold_local' ):
+            # threshold_local requires the additional parameter 'block_size'
+            def wrapper(im):
+                return im > func(im, 15)
+        elif ( func.__name__ == 'threshold_phansalkar' ):
+            # threshold_phansalkar requires an equalized image
+            def wrapper(im):
+                im_eq = equalize_adapthist(im)
+                return im_eq > func(im_eq)
+        else:
+            # standard case
+            def wrapper(im):
+                return im > func(im)
+
```
I don't know if you would accept this as good style, but it works for thresholding algorithms which require different input variables.

lagru replied:
Why not check for the actual function instead of a string comparison? E.g.
```python
if func is threshold_local:
    ...
```
Also the parenthesis and extra space around the conditional expressions are a bit unconventional. I'm surprised that @pep8speaks isn't bugging you about it.

kleinerELM replied:
Thanks for your advice, I changed it.
Actually, the parenthesis and extra space are a quirk of mine.

kleinerELM committed:
test for function instead of string

kleinerELM committed:
PEP 8 space issue

lagru commented on skimage/filters/thresholding.py:
```diff
@@ -117,10 +122,22 @@ def try_all_threshold(image, figsize=(8, 5), verbose=True):
+                return im > func(im)
+
         try:
             wrapper.__orifunc__ = func.__orifunc__
```
Am I correct in assuming that this line will always raise an AttributeError? At least that's what local testing tells me. Assuming you are the author, @sciunto do you still remember the intention behind this line?

I may be a good idea to use [functools.wraps](https://docs.python.org/3/library/functools.html?highlight=wraps#functools.wraps) instead; `wraps` preserves the name of the function and makes the original one accessible through the `__wrapped__` attribute.

kleinerELM replied:
Can you make a code suggestion? I just followed the existing pattern here. I wanted to avoid breaking anything.

lagru commented on skimage/filters/thresholding.py:
```diff
@@ -124,11 +124,11 @@ def thresh(func):
         Select a wrapper function to return a thresholded image. 
         """
-        if ( func.__name__ == 'threshold_local' ):
+        if func == threshold_local:
```
```suggestion
        if func is threshold_local:
```
We want to test for the identity of the function, not for its value. A semantic but in my opinion good distinction to make. :wink:

kleinerELM replied:
Oh, I didn't know the operator. I am learning a lot more than anticipated =)

lagru commented on skimage/filters/thresholding.py:
```diff
@@ -124,11 +124,11 @@ def thresh(func):
             def wrapper(im):
                 return im > func(im, 15)
-        elif ( func.__name__ == 'threshold_phansalkar' ):
+        elif func == threshold_phansalkar:
```
```suggestion
        elif func is threshold_phansalkar:
```

lagru committed:
Apply suggestions from code review

Compare functions with "is" instead of "==".

lagru commented:
@kleinerELM I took the liberty to commit two small suggestions. You should be able to update your local branch with `git pull origin threshold_phansalkar` (in case you didn't know).

matthew-brett commented:
We have changed the base branch for this PR from the `main` branch to the `pre-megamove-integration` branch.

## Summary - what this means for your PR

You can continue to work on your PR, now based on `pre-megamove-integration`.  You should update your branch to the current state of `pre-megamove-integration` (this is what the Github "Update branch" button does).   Please **do not merge the scikit-image `main` branch into your PR branch**.

Once we've merged your PR into the `pre-megamove-integration` branch, we'll update your changes for Megamove, and merge into the `main` branch.

In addition - our automated script detected that your branch had merge conflicts with the pre-Megamove branch. Please make sure to merge in the `pre-megamove-integration` branch, and resolve the merge conflicts in the way that you want.

Read on for the Megamove rationale, and more detailed instructions.

<details>

## Rationale

We've changed the base branch to make it easier and smoother to continue to work on your PR, while we work through a rather large change in the directory structure of the repository - a change that we have called *Megamove*.

See [the Megamove explanation issue](https://github.com/scikit-image/scikit-image/issues/8232) for more detail.

Changing the base branch means you do not see merge conflicts related to the *Megamove* change, and can continue to work with the repository in its original state, without worrying about porting those changes across to the new post-Megamove structure.

As you'll see in the linked issue, you can continue to work on your PR as before, and you can continue to use the Github GUI to "Update branch" - this updates from the new `pre-megamove-integration` branch.   However, there is one caveat: please, do not try and merge the `scikit-image` `main` branch into this one, but instead, keep track of the `pre-megamove-integration` branch.  The issue explains more.

To emphasize again, **do not try and merge the scikit-image `main` branch**.  You will get Megamove related merge conflicts that are rather tricky to resolve correctly.

When we've approved your PR, we'll merge into the `pre-megamove-integration` branch, and then process it so we can merge it into the `main` branch.   We'll let you know when we've merged your PR changes into `main`.

</details>

Any questions, please reply here.
